### ISSUE TITLE
MTP under pipeline parallelism (PP2): acceptance collapses 45.6% -> 8%, MTP is net-negative vs SPEC=off, and async scheduling corrupts output (3090 + A4000, sm86)

### ISSUE THREAD

**Issue by ming616** [syv-ai/HyperQwen#160]
**MTP under pipeline parallelism (PP2): acceptance collapses 45.6%→8% (np-independent), MTP is net-negative vs SPEC=off, and async scheduling outputs corrupted text — Qwen3.8-27B on 3090+A4000 (sm86)**

# Environment

- Repo: syv-ai recipe @ bae2023 (main), vLLM 0.28.0 base + recipe patch series
- Model: Qwen3.8-27B-INT4 (RedHatAI, compressed-tensors int4 sym g128; local head requant: lm_head/embed_tokens int8 g128, MTP linears int8 with mtp.fc kept bf16)
- Hardware: RTX 3090 24GB (GPU0) + RTX A4000 16GB (GPU1), PCIe, no P2P/NVLink between the two cards
- Serving: `--pipeline-parallel-size 2 --distributed-executor-backend mp`, `VLLM_PP_LAYER_PARTITION=48,16`, `--attention-backend TRITON_ATTN --kv-cache-dtype int4_per_token_head`, `INT8_ACT=int8` (W4A8 Marlin), `MAX_LEN=262144`, `MAX_SEQS=2/3/4`, `PREFIX_CACHE=1`, `GPU_UTIL=0.90`, `DRAFT_TOKENS=4`, `MTP_DRAFT_VOCAB=0`
- Baseline (TP2 `--tensor-parallel-size 2`, same model/flags): everything works — MTP acceptance 45.6%, needle 246K PASS

# Local patch required to even boot PP+MTP

Upstream `_pp_broadcast_prev_sampled_token_ids` asserts `sampled_token_ids.shape[-1] == 1`, which does not hold under speculative decoding. With a minimal local patch (broadcast `sampled_token_ids[:, :1].contiguous()`, receiver stays `[num_reqs, 1]`) the engine boots and serves. Without any patch, PP+MTP cannot start (shape assertion).

# Results (median of 3 runs each, 256-token generations, greedy, thinking off)

| Configuration | single tps | conc2 agg | conc4 agg | MTP acceptance | KV pool |
|---|---|---|---|---|---|
| TP2 + MTP (baseline) | 75.6 (EN) / 69.9 (CN) | 132.7 | — | 45.6% | 581,403 (2.22×) |
| PP2 sync + MTP (np2/3/4) | 19.7 / 19.8 / 19.8 | 48.6 | 72.3 (np4) | **7.5–8.4%** | 726K–792K (2.8–3.0×) |
| PP2 sync + SPEC=off (np2) | 32.9 | 61.4 | 55.6 | — | 778,942 (2.97×) |
| PP2 async + MTP (np2) | serves | — | — | — | — **output corrupted** |
| PP2 async + SPEC=off (np4) | 36.9 | 68.8 | **134.2** | — | **844,852 (3.22×)** |

Correctness spot-checks (needle-in-haystack, exact passphrase): 246K PASS on TP2+MTP and on PP2+async+SPEC=off; 32K PASS on both.

# Findings

1. **Acceptance collapse is np-independent and reproducible**: 45.6% (TP2) → 7.5–8.4% (PP2) across MAX_SEQS=2/3/4. With a controlled A/B against `SPEC=off` on the same PP2 setup, MTP under PP is a **net negative** (single-stream 32.9 → 19.8 tps, −40%): the broken drafter pays 4 chained draft forwards + a 5-token verify per step and gets almost everything rejected.
2. **Async + MTP corrupts output**: boot completes (boot-time "shared memory broadcast block" warnings appear but are benign), the first request returns degenerated repetition from token 1 (e.g. `"Hello!! How can can I helpHello How I"`). Related in spirit to #121 (async + garbage output on a different setup), but here it is specifically PP + spec.
3. **PP itself is healthy**: async + `SPEC=off` serves cleanly (needle PASS), gives a 3.22× KV pool (844,852 tokens) and ~49% faster prefill than TP2 (32K needle 15.9s vs 23.7s). The broken part is specifically the speculative-decoding plumbing across pipeline stages.

# Suspected causes

- **sync acceptance collapse**: drafter state under last-rank confinement looks broken — prime suspects are (a) the drafter's KV/attention metadata wiring on the last rank, and (b) token-embedding lookup for draft inputs: under PP the embedding table belongs to the first rank, so a last-rank-confined MTP head must obtain draft-token embeddings another way. The collapse being independent of MAX_SEQS suggests a systematic input corruption to the drafter rather than a scheduling artifact.
- **async corruption**: broadcasting only `sampled_token_ids[:, :1]` is not a sufficient contract for the async spec path — the draft-chain continuation needs the accepted-token set (or acceptance metadata) across stages, not just one column. The proper fix is probably a spec-aware PP broadcast contract (accepted prefix + lengths), i.e. draft-model `SupportsPP`-grade work.

# Question

Is PP + MTP within scope for this recipe, and is there guidance on what the intended PP contract for the drafter is (broadcast payload, drafter KV/embedding placement)? If it is out of scope, we can run PP in `SPEC=off` mode (which works well: 4 slots, 3.22× pool, conc4 134 tok/s agg) — but it would be good to document that combination explicitly, since the failure mode of MTP under PP (8% acceptance, net-negative vs no-spec) is easy to walk into unknowingly.

Happy to provide full boot logs, tracebacks, and repro scripts.

---
Engine-level companion report (same investigation, broadcast contract + drafter PP semantics): https://github.com/vllm-project/vllm/issues/57755

**Comment by mhenrichsen** [comment]
Retitled from "Title" — the real one was in the body.

This is a good report, and the source backs your hypothesis (a) over (b). Checked against a pristine `v0.28.0` checkout:

**The drafter is not declared PP-capable at all.** `Qwen3_5MTP` in `vllm/model_executor/models/qwen3_5_mtp.py:211` is

```python
class Qwen3_5MTP(LocalArgmaxMixin, nn.Module, SupportsMultiModal):
```

— no `SupportsPP`. Four other MTP implementations in the same release do declare it (`bailing_moe_v3_mtp`, `glm_ocr_mtp`, `glm4_moe_lite_mtp`, `nemotron_h_mtp`), so upstream does treat that interface as the gate for "this draft model may be split across pipeline stages", and this family has not been through it. Your closing guess — that the proper fix is draft-model `SupportsPP`-grade work — is the right read.

**Hypothesis (b), the embedding, is probably not it.** The MTP module builds its own `VocabParallelEmbedding` (`qwen3_5_mtp.py:82`, used at line 137) and pulls `embed_tokens`/`lm_head` out of the weight stream by name (lines 305-306), so the drafter has a local embedding table wherever the module lands. Worth confirming on your box that those tensors actually loaded on rank 1 rather than being silently skipped by the PP weight filter — but the design does not require reaching back to rank 0 for them.

**The async corruption is explained by the contract, exactly as you say.** `_pp_broadcast_prev_sampled_token_ids` (`vllm/v1/worker/gpu_model_runner.py:4951`) asserts `[num_reqs, 1]`, and the receiving side allocates `torch.empty((num_reqs, 1))` and stores it as `input_batch.prev_sampled_token_ids` for `_prepare_input_ids`. Under speculation the previous step accepts a variable 1..k+1 tokens per request, so one column cannot reconstruct what the earlier ranks must feed — your `[:, :1]` patch makes it boot and then hands rank 0 a wrong prefix, which is precisely the degenerate repetition you see from token 1. That is a contract gap, not a bug you can patch at the broadcast site; it needs accepted-prefix + lengths.

**On scope:** yes, out of scope for this recipe. Everything published here is single-card or tensor-parallel; `docs/multi-gpu.md` covers the TP shape and the layer/KV-head arithmetic, and no profile in the repo sets `--pipeline-parallel-size`. So PP + MTP is not a regression against anything we ship — but I agree the failure mode is easy to walk into, and your A/B is the useful half of that: **PP2 + `SPEC=off` is healthy** (needle 246K PASS, 3.22x KV pool, 134 tok/s aggregate at conc4, ~49% faster prefill than TP2), while PP2 + MTP is a net loss against no speculation at all.

If you are willing, I would like to put that in `docs/multi-gpu.md` as a short "pipeline parallelism" note — the two cards, `VLLM_PP_LAYER_PARTITION=48,16`, `SPEC=off`, the pool and needle results, and the one-line warning that MTP under PP collapses to ~8% acceptance. Say the word and whether you want it attributed, and confirm: was the TP2 baseline at the same `MAX_LEN` and `MAX_SEQS` as the PP2 rows, and did you cap both cards at the same power limit? The A4000 is a 140 W part next to a 350 W 3090, which is fine for a PP A/B (the ranks run in sequence) but worth stating in the note.

Related: [#121](https://github.com/syv-ai/HyperQwen/issues/121) is async + garbage output on a 4090 with MTP3 + FP8, where the workaround is also `--no-async-scheduling`. Different mechanism, likely, but if the async path is fragile in two unrelated configurations that is worth knowing. And thank you for filing the engine-level half upstream (vllm-project/vllm#57755) — that is where the contract change has to happen.
